src/miner.cpp

```diff
@@ -351,332 +385,387 @@ CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const MinerAddre
-    bool nonEmptyBlock = !next_cb_mtx;
+    blockFinished = blockFinished && next_cb_mtx;
 
-    // Largest block you're willing to create:
-    unsigned int nBlockMaxSize = GetArg("-blockmaxsize", DEFAULT_BLOCK_MAX_SIZE);
-    // Limit to between 1K and MAX_BLOCK_SIZE-1K for sanity:
-    nBlockMaxSize = std::max((unsigned int)1000, std::min((unsigned int)(MAX_BLOCK_SIZE-1000), nBlockMaxSize));
+    LOCK2(cs_main, mempool.cs);
+    CBlockIndex* pindexPrev = chainActive.Tip();
+    nHeight = pindexPrev->nHeight + 1;
+    uint32_t consensusBranchId = CurrentEpochBranchId(nHeight, chainparams.GetConsensus());
 
-    // How much of the block should be dedicated to high-priority transactions,
-    // included regardless of the fees they pay
-    unsigned int nBlockPrioritySize = GetArg("-blockprioritysize", DEFAULT_BLOCK_PRIORITY_SIZE);
-    nBlockPrioritySize = std::min(nBlockMaxSize, nBlockPrioritySize);
+    // -regtest only: allow overriding block.nVersion with
+    // -blockversion=N to test forking scenarios
+    if (chainparams.MineBlocksOnDemand())
+        pblock->nVersion = GetArg("-blockversion", pblock->nVersion);
 
-    // Minimum block size you want to create; block will be filled with free transactions
-    // until there are no more or the block reaches this size:
-    unsigned int nBlockMinSize = GetArg("-blockminsize", DEFAULT_BLOCK_MIN_SIZE);
-    nBlockMinSize = std::min(nBlockMaxSize, nBlockMinSize);
+    pblock->nTime = GetTime();
+    const int64_t nMedianTimePast = pindexPrev->GetMedianTimePast();
+    CCoinsViewCache view(pcoinsTip);
+
+    SaplingMerkleTree sapling_tree;
+    assert(view.GetSaplingAnchorAt(view.GetBestAnchor(SAPLING), sapling_tree));
+
+    nLockTimeCutoff = (STANDARD_LOCKTIME_VERIFY_FLAGS & LOCKTIME_MEDIAN_TIME_PAST)
+                       ? nMedianTimePast
+                       : pblock->GetBlockTime();
+
+    // We want to track the value pool, but if the miner gets
+    // invoked on an old block before the hardcoded fallback
+    // is active we don't want to trip up any assertions. So,
+    // we only adhere to the turnstile (as a miner) if we
+    // actually have all of the information necessary to do
+    // so.
+    if (chainparams.ZIP209Enabled()) {
+        if (pindexPrev->nChainSproutValue) {
+            sproutValue = *pindexPrev->nChainSproutValue;
+        } else {
+            monitoring_pool_balances = false;
+        }
+        if (pindexPrev->nChainSaplingValue) {
+            saplingValue = *pindexPrev->nChainSaplingValue;
+        } else {
+            monitoring_pool_balances = false;
+        }
+        if (pindexPrev->nChainOrchardValue) {
+            orchardValue = *pindexPrev->nChainOrchardValue;
+        } else {
+            monitoring_pool_balances = false;
+        }
+    }
 
-    // Collect memory pool transactions into the block
-    CTxMemPool::setEntries inBlock;
-    CTxMemPool::setEntries waitSet;
+    addPriorityTxs();
+    addScoreTxs();
```

Comment:
When we get to the ZIP 317 implementation, it'd be nice to just make this whole thing into a function instead of using a pile of mutable state to perform an operation that can essentially be pure.